org.osgi.test.common/src/main/java/org/osgi/test/common/annotation/config/WithConfigurationUtil.java

```diff
@@ -0,0 +1,56 @@
+package org.osgi.test.common.annotation.config;
+
+import java.util.UUID;
+
+public class WithConfigurationUtil {
+
+	private WithConfigurationUtil() {
+
+	}
+
+	public static String pid(WithConfiguration withConfiguration) {
+		return pid(withConfiguration.pid());
+	}
+
+	public static String pid(InjectConfiguration injectConfiguration) {
+		return pid(injectConfiguration.value());
+	}
+
+	public static String pid(String pid) {
+		if (isFactory(pid)) {
+			return pid.substring(0, pid.indexOf("~"));
+		} else {
+			return pid;
+		}
+	}
+
+	public static String factoryName(WithConfiguration withConfiguration) {
+		return factoryName(withConfiguration.pid());
+	}
+
+	public static String factoryName(InjectConfiguration injectConfiguration) {
+		return factoryName(injectConfiguration.value());
+	}
+
+	public static String factoryName(String pid) {
+		if (!isFactory(pid)) {
+			throw new IllegalArgumentException("It is not a factory. ~ is missing");
+		}
+		String tmp = pid.substring(pid.indexOf("~") + 1);
+
+		return tmp.isEmpty() ? UUID.randomUUID()
+			.toString() : tmp;
+	}
+
+	public static boolean isFactory(WithConfiguration withConfiguration) {
+		return isFactory(withConfiguration.pid());
+	}
+
+	public static boolean isFactory(InjectConfiguration injectConfiguration) {
+		return isFactory(injectConfiguration.value());
+	}
+
+	public static boolean isFactory(String pid) {
+		return pid.contains("~");
```

Comment:
Why does a factory configuration require `~`? This is just a convention for ConfigAdmin to generate a configuration pid from a factory pid and a name.

Reply:
removed most of the "~" magic